controllers/policies/admissionpolicy_gc_controller.go

```diff
@@ -0,0 +1,95 @@
+package policies
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	policiesv1alpha2 "github.com/kubewarden/kubewarden-controller/apis/policies/v1alpha2"
+	"github.com/kubewarden/kubewarden-controller/internal/pkg/admission"
+	"github.com/kubewarden/kubewarden-controller/internal/pkg/constants"
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// AdmissionPolicyGCReconciler deletes namespaced policies that are no longer
+// attached to any policy server
+type AdmissionPolicyGCReconciler struct {
+	client.Client
+	Log        logr.Logger
+	Scheme     *runtime.Scheme
+	Reconciler admission.Reconciler
+}
+
+func (r *AdmissionPolicyGCReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	admissionPolicy := policiesv1alpha2.AdmissionPolicy{}
+	if err := r.Get(ctx, req.NamespacedName, &admissionPolicy); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("cannot retrieve admission policy: %w", err)
+	}
+	policyServerWasDeleted := false
+	if admissionPolicy.Spec.PolicyServer != "" {
+		policyServer := policiesv1alpha2.PolicyServer{}
+		if err := r.Get(ctx, types.NamespacedName{Name: admissionPolicy.Spec.PolicyServer}, &policyServer); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return ctrl.Result{}, errors.Wrapf(err, "could not get policy server %s", admissionPolicy.Spec.PolicyServer)
+			}
+		} else {
+			policyServerWasDeleted = policyServer.DeletionTimestamp != nil
+		}
+	}
+	if admissionPolicy.DeletionTimestamp == nil {
+		if policyServerWasDeleted {
+			err := r.Delete(ctx, &admissionPolicy)
+			if err == nil {
+				return ctrl.Result{}, nil
+			} else if !apierrors.IsNotFound(err) {
+				return ctrl.Result{}, errors.Wrap(err, "could not delete admission policy")
+			}
+		} else {
+			return ctrl.Result{}, nil
+		}
+	}
+	controllerutil.RemoveFinalizer(&admissionPolicy, constants.KubewardenFinalizer)
+	if err := r.Update(context.TODO(), &admissionPolicy); err != nil {
+		return ctrl.Result{}, fmt.Errorf("cannot update admission policy: %w", err)
+	}
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *AdmissionPolicyGCReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	err := ctrl.NewControllerManagedBy(mgr).
+		For(&policiesv1alpha2.AdmissionPolicy{}).
```

Comment:
is it possible to add `For(&policiesv1alpha2.ClusterAdmissionPolicy{}).` here too? Then we could get all the policies (both AdmissionPolicy and ClusterAdmissionPolicy) and reconcile them. This would reduce the duplicated code, but I'm not sure if that's even possible

Reply:
it seems that it's not possible to do it